crypto/pkcs7/pkcs7_x509.c

```diff
@@ -378,27 +378,25 @@ static int write_signer_info(CBB *out, const void *arg) {
 }
 
 static int pkcs7_add_signature(PKCS7 *p7, X509 *x509, EVP_PKEY *pkey) {
+  // OpenSSL's docs say that this defaults to SHA1, but appears to actually
+  // default to SHA256 in 1.1.x and 3.x for RSA, DSA, and EC(DSA).
+  // https://linux.die.net/man/3/pkcs7_sign
+  // https://github.com/openssl/openssl/blob/79c98fc6ccab49f02528e06cc046ac61f841a753/crypto/rsa/rsa_ameth.c#L438
+  const EVP_MD *digest = EVP_sha256();
   PKCS7_SIGNER_INFO *si = NULL;
-  const EVP_MD *digest = NULL;
 
-  EVP_PKEY_CTX *pkey_ctx = EVP_PKEY_CTX_new(pkey, /*engine*/ NULL);
-  if (!pkey_ctx) {
-    goto err;
-  }
-  int ok = EVP_PKEY_CTX_get_signature_md(pkey_ctx, &digest);
-  EVP_PKEY_CTX_free(pkey_ctx);
-  if (!ok && (EVP_PKEY_id(pkey) == EVP_PKEY_RSA ||
-              EVP_PKEY_id(pkey) == EVP_PKEY_DSA)) {
-    // OpenSSL's docs say that this defaults to SHA1, but appears to actually
-    // default to SHA256 in 1.1.x and 3.x
-    // https://linux.die.net/man/3/pkcs7_sign
-    // https://github.com/openssl/openssl/blob/79c98fc6ccab49f02528e06cc046ac61f841a753/crypto/rsa/rsa_ameth.c#L438
-    digest = EVP_sha256();
-  } else if (!ok) {
-    OPENSSL_PUT_ERROR(PKCS7, PKCS7_R_NO_DEFAULT_DIGEST);
-    goto err;
+  switch (EVP_PKEY_id(pkey)) {
+    case EVP_PKEY_RSA:
+    case EVP_PKEY_DSA:
+    case EVP_PKEY_EC:
+      break;
+    default:
+      OPENSSL_PUT_ERROR(PKCS7, PKCS7_R_NO_DEFAULT_DIGEST);
+      goto err;
```

Comment:
The new logic can only use SHA256, with no alternatives. The previous logic honored a digest if one was defaulted to in `EVP_PKEY_CTX_new` -- should we continue to do this?

Reply:
Hm, how is that `EVP_PKEY_CTX_new` default configured? I don't see anything that would do that in [`EVP_PKEY_CTX_new`]() or e.g. `

Reply:
Some additional context here is that we've adapted some public OpenSSL functions, namely [`PKCS7_sign_add_signer`](https://github.com/openssl/openssl/blob/5fce85ec52a826d53665552b50e67f86c92dc394/crypto/pkcs7/pk7_smime.c#L118) and [`PKCS7_add_signature`](https://github.com/openssl/openssl/blob/5fce85ec52a826d53665552b50e67f86c92dc394/crypto/pkcs7/pk7_lib.c#L393), into internal static functions and have pared them down a bit. as external functions, each of those take a `const EVP_MD *dgst` parameter, allowing the caller to specify which digest algorithm is used in computing the signature.

However, I noticed that `PKCS7_sign`'s [invocation of `PKCS7_sign_add_signer`](https://github.com/openssl/openssl/blob/5fce85ec52a826d53665552b50e67f86c92dc394/crypto/pkcs7/pk7_smime.c#L42) unconditionally `NULL`s out the `md` parameter, meaning that the default is always taken for signers added in the one-shot call to `PKCS7_sign`.

I suppose OpenSSL users can configure non-default digest algorithms, but because we (intentionally) don't externalize `PKCS7_sign_add_signer` and `PKCS7_add_signature` AWS-LC users cannot. This comports with our prior prescriptivism in "mode 2" of PKCS7_sign where SHA256 is used. If future integration projects require a configurable digest, we can always expand our API surface area to accommodate.